// docusaurus-theme-redoc/dist/theme/Redoc/Redoc.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importDefault(require("react"));
const useGlobalData_1 = require("@docusaurus/useGlobalData");
const useThemeContext_1 = __importDefault(require("@theme/hooks/useThemeContext"));
const redoc_1 = require("redoc");
require("./styles.css");
function Redoc(props) {
    const { isDarkTheme } = useThemeContext_1.default();
    const { lightTheme, darkTheme, redocOptions = null, } = useGlobalData_1.usePluginData('docusaurus-theme-redoc');
    const theme = isDarkTheme ? darkTheme : lightTheme;
    const { spec, specUrl } = props;
    return (react_1.default.createElement("div", { className: "redocusaurus" },
        react_1.default.createElement(redoc_1.RedocStandalone, { spec: spec, specUrl: specUrl, options: {
                scrollYOffset: 'nav.navbar',
                hideDownloadButton: true,
                expandSingleSchemaField: true,
                menuToggle: true,
                // @ts-expect-error not in types
                suppressWarnings: true,
                ...redocOptions,
                theme,
            } })));
}
exports.default = Redoc;
